JP-3848 Update S_REGION for MIRI LRS fixed slit data

jemorrison opened the pull request:
Closing this and replacing it with https://github.com/spacetelescope/jwst/pull/9193

<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3848](https://jira.stsci.edu/browse/JP-3848)

This PR needs a new datamodel that defines MIRI LRS specwcs reference files. The stdatamodels for this change is
https://github.com/spacetelescope/stdatamodels/pull/383. The datamodel PR must be merged before this PR can be
merged. 


THIS PR IS NOT READY TO BE MERGED. We still need to make changes to the resample code when it updates the S_REGION for MIRI LRS data. 

This PR is not correct for updating the s-region for spec 3 MIRI LRS data. Still testing code. 

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #

<!-- describe the changes comprising this PR here -->
This PR correctly updates the s_region set by assign_wcs for MIRI LRS fixed slit data. The change requires using the slit corners in V2,V3 space which were added to the MIRI LRS specwcs reference file. 

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] **request a review from someone specific**, to avoid making the maintainers review every PR
- [ ] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

nden commented on jwst/assign_wcs/miri.py:
```diff
@@ -253,19 +251,14 @@ def lrs_xytoabl(input_model, reference_files):
-            zero_point = ref[0].header['imxsltl'] - 1, ref[0].header['imysltl'] - 1
-            # Transform to slitless subarray from full array
-            zero_point = subarray2full.inverse(zero_point[0], zero_point[1])
+    refmodel = MIRILrsModel(reference_files['specwcs'])
```
`refmodel` will need to be closed at the end

nden commented on jwst/assign_wcs/miri.py:
```diff
@@ -405,18 +398,14 @@ def lrs_abltov2v3l(input_model, reference_files):
-            zero_point = ref[0].header['imxsltl'] - 1, ref[0].header['imysltl'] - 1
-            # Transform to slitless subarray from full array
-            zero_point = subarray2full.inverse(zero_point[0], zero_point[1])
+    refmodel = MIRILrsModel(reference_files['specwcs'])
```
close the file at the end

nden commented on jwst/assign_wcs/util.py:
```diff
@@ -807,6 +807,37 @@ def update_s_region_imaging(model):
+    Update the ``S_REGION`` keyword using V2,V3 of the slit corners from reference file`.
+    """
+
+    refmodel = MIRILrsModel(reference_files['specwcs'])
```
close the file

nden reviewed (changes requested):
- The code as written does not allow using the old file. Is this desirable to make the transition easier?

- Line [317 ](https://github.com/spacetelescope/jwst/blob/main/jwst/assign_wcs/miri.py#L317-L318) was added to account for the fact that the Tabular model defines its own range for inputs. If the bounding_box is larger then s_region will be filled with NaNs because this is the fill_value for the Tabular model. It's unlikely the slit corners are outside the range of the Tabular model so those lines should probably be removed.

- Also this needs some kind of test. It may be an addition to an existing test to check the `s_region` has a reasonable value.

jemorrison commented on jwst/assign_wcs/util.py:
```diff
@@ -807,6 +807,37 @@ def update_s_region_imaging(model):
 
+def update_s_region_lrs(model, reference_files):
+    """
+    Update the ``S_REGION`` keyword using V2,V3 of the slit corners from reference file`.
```
@nden I need a clarification on not needing dxmodel (line 317 any more)
Does that mean that we no longer need to determine bb_sub in line 311. So lines 301 to 311 can me cut:
These lines:
 # Fit for X shift as a function of Y
    # Spline fit with enforced smoothness
    spl = UnivariateSpline(ycen_subarray[::-1], xcen_subarray[::-1] - zero_point[0], s=0.002)
    # Evaluate the fit at the y reference points
    xshiftref = spl(ycen_subarray)
    # This function will give slit dX as a function of Y subarray pixel value
    dxmodel = models.Tabular1D(lookup_table=xshiftref, points=ycen_subarray, name='xshiftref',
                                 bounds_error=False, fill_value=np.nan)

    if input_model.meta.exposure.type.lower() == 'mir_lrs-fixedslit':
        bb_sub = (bb_sub[0], (dxmodel.points[0].min(), dxmodel.points[0].max()))

jemorrison commented:
@nden If I remove following in miri.py 
   # dxmodel = models.Tabular1D(lookup_table=xshiftref, points=ycen_subarray, name='xshiftref',
    #                             bounds_error=False, fill_value=np.nan)

Then later on in the code dxmodel is used.  How should xmodel be defined:

 # What is the effective slit X as a function of subarray x,y?
    xmodel = models.Mapping([0], n_inputs=2) - (models.Mapping([1], n_inputs=2) | dxmodel)

jemorrison commented:
@nden 
You were concerned:
"As the code as written does not allow using the old file. Is this desirable to make the transition easier?'
As it is written if it uses the old reference file the v2 v3 values are set to NaN and we get the message:
025-01-31 14:43:49,200 - stpipe.AssignWcsStep - INFO - There are NaNs in s_region, S_REGION not updated.

So the s_region is not updated and it is just what set telescope pointing set which I think is more desirable that the what the
old code did and updated the s_region to very incorrect values.
Is this acceptable to you ? I have also added an INFO message that the V2,V3 corners in the reference file are missing and signaling that no updated s_region will result.

jemorrison commented:
@nden 
I am working on a test. I would like to set up an LRS data model using a hard coded set of wcs values and s_region.
I can also set up the lrspec wcs reference file. 
Then call the update_s_region_lrs in util.py - that needs the v2v3 transform to world and test that new s_region similar to the original s_region (diff less than some tolerance)
I am just not sure how to set up the wcs of the lrs data.  When I look at the wcs of a cal lrs miri file - well it contain A LOT of info. Could you give me a suggestion on how to get started ?

nden commented:
@jemorrison regarding setting up the WCS of a LRS cal file - could you create in memory ImageModel with the appropriate configuration and call AssignWcs on it. Alternativelly add a check for s_region in one of the existing regression tests. May be [this one](https://github.com/spacetelescope/jwst/blob/main/jwst/regtest/test_miri_lrs_slit_spec2.py#L68) will work.

I agree with the decision about the old file.

Also, I suggest you remove the lines below, not dxmodel. These reset the bounding_box and will have an effect on computing s_region.

```
if input_model.meta.exposure.type.lower() == 'mir_lrs-fixedslit':
    bb_sub = (bb_sub[0], (dxmodel.points[0].min(), dxmodel.points[0].max()))
```

jemorrison commented:
@nden What I have written in resample_utils.py combine_s_regions_lrs is not correct.
This code will depend on the PA. I did not want to change the wcs for the lrs in resample_spec since the data is being combined correctly - all that is wrong is the final s_region values in the header.
Maybe I can remove the PA then find the combined_s_region using the code I have (or something similar) and then apply the PA to the footprint values. 
Suggestions ?

drlaw1558 commented:
@jemorrison I'm having trouble running this PR myself (I pointed to your MIRI_LRS or MIRI_LRS_SpecWCS branches of stdatamodels, but I'm still getting odd-looking results).  The plot above looks good though.

Can we not construct the S_REGION of the output result from the S_REGION of all of the individual input files?  Or do we not have access to that information any longer at the stage that we're trying to compute S_REGION for the resampled data?  Alternatively, perhaps we could use one of the input S_REGIONs to compute the slit width dynamically (compute distance on sky from corner 1 to corners 2/3/4 and pick the minimum distance), and then use that width with the central line that you're getting from the bounding box approach above after resampling?  That way we'd avoid hardcoding a value for the slit width.

jemorrison commented:
@nden How can I combine the s_region from two cal files to get the final s_region that covers both s_regions ?

emolter commented:
> How can I combine the s_region from two cal files to get the final s_region that covers both s_regions ?

I think you're looking for `wcs_from_sregions` in `stcal`: https://github.com/spacetelescope/stcal/blob/b35bccc823c2faa9df40bc05999763c6a5b10995/src/stcal/alignment/util.py#L675

jemorrison commented:
I looked at wcs_from_sregions I I don't think I want that. I have a wcs of a slit but it only gives the ra,dec of slit center. But I have the s_regions from each cal file that is  correct s_region for that exposure. I want to use only the two s_regions and somehow combine them and find a final s_region. Looking at astropy regions and I think I do this somehow.

jemorrison commented:
Closing this PR in favor of https://github.com/spacetelescope/jwst/pull/9193
